File: android/src/main/kotlin/com/khanhtq/app_blocker/event/BlockEventStreamHandler.kt
package com.khanhtq.app_blocker.event

import android.os.Handler
import android.os.Looper
import io.flutter.plugin.common.EventChannel

/**
 * Singleton [EventChannel.StreamHandler] that delivers block events to Flutter.
 *
 * Any component — including [com.khanhtq.app_blocker.blocking.AppBlockerAccessibilityService]
 * and the scheduling receivers — can call [sendEvent] from any thread. Events
 * are always posted to the main looper before being forwarded to the Flutter
 * engine, which is the only thread on which [EventChannel.EventSink.success]
 * may be called safely.
 *
 * ### Calling conventions
 *
 * **Instance-based** (from the plugin, where the instance is readily accessible):
 * ```kotlin
 * BlockEventStreamHandler.instance.sendEvent("blocked", packageName = "com.example.app")
 * ```
 *
 * **Static map-based** (from services and receivers that only have access to
 * the companion object via a static import):
 * ```kotlin
 * BlockEventStreamHandler.sendEvent(
 *     mapOf("type" to "blocked", "packageName" to "com.example.app", "timestamp" to millis)
 * )
 * ```
 */
class BlockEventStreamHandler private constructor() : EventChannel.StreamHandler {

    companion object {
        /** Shared instance registered with the [EventChannel]. */
        val instance: BlockEventStreamHandler by lazy { BlockEventStreamHandler() }

        private val mainHandler = Handler(Looper.getMainLooper())

        /**
         * Posts [event] to the Flutter sink on the main thread.
         * Safe to call from any thread. No-ops if no listener is attached.
         */
        @JvmStatic
        fun sendEvent(event: Map<String, Any?>) {
            // Snapshot the sink reference once to avoid TOCTOU races.
            val sink = instance.eventSink ?: return

            if (Looper.myLooper() == Looper.getMainLooper()) {
                sink.success(event)
            } else {
                mainHandler.post { sink.success(event) }
            }
        }
    }

    /**
     * The active Flutter event sink.
     *
     * Marked [@Volatile] so that reads in [sendEvent] (potentially on a
     * background thread) always observe the most recent write from the main
     * thread inside [onListen] / [onCancel].
     */
    @Volatile
    private var eventSink: EventChannel.EventSink? = null

    // ------------------------------------------------------------------
    // StreamHandler
    // ------------------------------------------------------------------

    override fun onListen(arguments: Any?, events: EventChannel.EventSink?) {
        eventSink = events
    }

    override fun onCancel(arguments: Any?) {
        eventSink = null
    }

    // ------------------------------------------------------------------
    // Instance API
    // ------------------------------------------------------------------

    /**
     * Sends a typed block event to Flutter.
     *
     * @param type        Event type matching Dart's `BlockEventType` values:
     *                    `"blocked"`, `"unblocked"`, `"attemptedAccess"`,
     *                    `"scheduleActivated"`, `"scheduleDeactivated"`.
     * @param packageName Package name of the affected app, or `null`.
     * @param scheduleId  ID of the schedule that triggered the event, or `null`.
     */
    fun sendEvent(type: String, packageName: String? = null, scheduleId: String? = null) {
        sendEvent(
            buildMap {
                put("type", type)
                put("timestamp", System.currentTimeMillis())
                if (packageName != null) put("packageName", packageName)
                if (scheduleId != null) put("scheduleId", scheduleId)
            }
        )
    }
}

File: ios/Classes/BlockEventStreamHandler.swift
import Foundation
import Flutter

/// Delivers block events from native iOS code to the Flutter event stream.
///
/// `FlutterEventSink` must only be called on the main thread. All calls to
/// `sendEvent` dispatch to `DispatchQueue.main` before invoking the sink, so
/// callers may call from any thread.
class BlockEventStreamHandler: NSObject, FlutterStreamHandler {

    /// The active event sink, set by Flutter when a listener subscribes.
    /// Accessed only on the main thread (via `DispatchQueue.main.async`).
    private var eventSink: FlutterEventSink?

    // MARK: - FlutterStreamHandler

    func onListen(withArguments arguments: Any?, eventSink events: @escaping FlutterEventSink) -> FlutterError? {
        eventSink = events
        return nil
    }

    func onCancel(withArguments arguments: Any?) -> FlutterError? {
        eventSink = nil
        return nil
    }

    // MARK: - Public API

    /// Sends a typed block event to the Flutter side.
    ///
    /// Safe to call from any thread.
    ///
    /// - Parameters:
    ///   - type: Event type matching Dart's `BlockEventType` values:
    ///           `"blocked"`, `"unblocked"`, `"attemptedAccess"`,
    ///           `"scheduleActivated"`, `"scheduleDeactivated"`,
    ///           `"profileActivated"`, `"profileDeactivated"`.
    ///   - packageName: Identifier of the affected app, or `nil`.
    ///   - scheduleId: ID of the schedule that triggered the event, or `nil`.
    ///   - profileId: ID of the profile that triggered the event, or `nil`.
    func sendEvent(type: String, packageName: String?, scheduleId: String?, profileId: String? = nil) {
        var event: [String: Any] = [
            "type": type,
            "timestamp": Int(Date().timeIntervalSince1970 * 1000),
        ]
        if let packageName = packageName { event["packageName"] = packageName }
        if let scheduleId = scheduleId { event["scheduleId"] = scheduleId }
        if let profileId = profileId { event["profileId"] = profileId }

        DispatchQueue.main.async { [weak self] in
            self?.eventSink?(event)
        }
    }
}
